File: 02/day2.erl
%% -*- erlang -*-
%% Usage: `cat ./tests/test3.txt | escript day2.erl`

%% Convert string to list of ints.
row_to_int_cols(String) ->
    StrCols = string:lexemes(String, "\s"),
    lists:map(fun(X) -> {Int, _} = string:to_integer(X), Int end, StrCols).

%% Use read_stdin/1. Read stdin until EOF and return as list of strings.
read_line(Rows) ->
    GetLine = io:get_line(""),
    if
        GetLine /= eof ->
            Line = string:trim(GetLine, trailing, "\n"),
            Cols = row_to_int_cols(Line),
            read_line([Cols] ++ Rows);
        true ->
            Rows
    end.
read_stdin() -> read_line([]).

%% Use row_max/1. Reduces row to max value of its columns.
row_max([Tail], Result) ->
    if
        Tail > Result ->
            Tail;
        true ->
            Result
    end;
row_max([Head | Rest], Result) ->
    if
        Head > Result ->
            row_max(Rest, Head);
        true ->
            row_max(Rest, Result)
    end.
row_max([Head | Rest]) -> row_max(Rest, Head).

%% Use row_min/1. Reduces row to min value of its columns.
row_min([Tail], Result) ->
    if
        Tail < Result ->
            Tail;
        true ->
            Result
    end;
row_min([Head | Rest], Result) ->
    if
        Head < Result ->
            row_min(Rest, Head);
        true ->
            row_min(Rest, Result)
    end.
row_min([Head | Rest]) -> row_min(Rest, Head).

%% Part 1 solution:
find_diff(Cols) ->
    Max = row_max(Cols),
    Min = row_min(Cols),
    erlang:abs(Max - Min).

checksum_one([Cols], Sum) ->
    Diff = find_diff(Cols),
    Sum + Diff;
checksum_one([Cols | RestRows], Sum) ->
    Diff = find_diff(Cols),
    checksum_one(RestRows, Sum + Diff).
checksum_one(Rows) -> checksum_one(Rows, 0).

%% Part 2 solution:
find_div_in_col(Num, [Tail]) ->
    if
        Num rem Tail == 0 ->
            Num / Tail;
        Tail rem Num == 0 ->
            Tail / Num;
        true ->
            none
    end;
find_div_in_col(Num, [Head | Rest]) ->
    if
        Num rem Head == 0 ->
            Num / Head;
        Head rem Num == 0 ->
            Head / Num;
        true ->
            find_div_in_col(Num, Rest)
    end.
find_div([_]) -> 0;
find_div([Head | Rest]) ->
    Result = find_div_in_col(Head, Rest),
    if
        Result == none ->
            find_div(Rest);
        true ->
            trunc(Result)
    end.

checksum_two([Cols], Sum) ->
    Div = find_div(Cols),
    Sum + Div;
checksum_two([Cols | RestRows], Sum) ->
    Div = find_div(Cols),
    checksum_two(RestRows, Sum + Div).
checksum_two(Rows) -> checksum_two(Rows, 0).

%% main/1 for escript.
main(_) ->
    Rows = read_stdin(),
    ChecksumOne = checksum_one(Rows),
    ChecksumTwo = checksum_two(Rows),
    erlang:display(ChecksumOne),
    erlang:display(ChecksumTwo).

File: 02/day2.rs
use std::io;

fn split_str_to_u32_vec(input: &str, output: &mut Vec<u32>) {
    // Turn a row slice into columns of integers.
    for token in input.split_whitespace() {
        let i: u32 = token.parse().unwrap();
        output.push(i);
    }
}

fn chksum_one(rows: &Vec<Vec<u32>>) -> u32 {
    let mut sum = 0;
    for row in rows {
        let max = *row.iter().max().unwrap() as i32;
        let min = *row.iter().min().unwrap() as i32;
        let diff = (max - min).abs() as u32;
        sum += diff;
    }

    sum
}

fn find_div(row: &Vec<u32>) -> u32 {
    let mut div = 0;
    for (i1, c1) in row.iter().enumerate() {
        for (i2, c2) in row.iter().enumerate() {
            if i1 == i2 {
                break;
            }
            if c1 % c2 == 0 {
                div = c1 / c2;
            } else if c2 % c1 == 0 {
                div = c2 / c1;
            }
        }
    }

    div
}

fn chksum_two(rows: &Vec<Vec<u32>>) -> u32 {
    let mut sum = 0;
    for row in rows {
        let div = find_div(&row);
        sum += div;
    }

    sum
}

fn main() {
    let mut rows: Vec<Vec<u32>> = vec![];
    let stdin = io::stdin();
    loop {
        let mut buffer = String::new();
        match stdin.read_line(&mut buffer) {
            Err(_) => {},
            Ok(bytes) => if bytes == 0 { break; }
        }
        let trimmed = String::from(buffer.trim_matches('\n'));
        let mut cols = vec![];
        split_str_to_u32_vec(&trimmed, &mut cols);
        &rows.push(cols);
    }

    println!("Part 1: {}", chksum_one(&rows));
    println!("Part 2: {}", chksum_two(&rows));
}
